R/module_teal_data.R

```diff
@@ -0,0 +1,203 @@
+      if (inherits(data_out, "shiny.silent.error") && identical(data_out$message, "")) {
+        if (!validate_shiny_silent_error) {
+          return(NULL)
+        } else {
+          validate(
+            need(
+              FALSE,
+              paste(
+                strip_style(data_out$message),
+                "Check your inputs or contact app developer if error persists.",
+                sep = ifelse(identical(data_out$message, ""), "", "\n")
+              )
+            )
+          )
+        }
+      }
+
+      # to handle errors and qenv.error(s)
+      if (inherits(data_out, c("qenv.error", "error"))) {
+        validate(
+          need(
+            FALSE,
+            paste(
+              "Error when executing `teal_data_module` passed to `data`:\n ",
+              strip_style(paste(data_out$message, collapse = "\n")),
+              "\n Check your inputs or contact app developer if error persists."
+            )
+          )
+        )
+      }
+
+      validate(
+        need(
+          inherits(data_out, "teal_data"),
+          paste(
+            "Error: `teal_data_module` passed to `data` failed to return `teal_data` object, returned",
+            strip_style(toString(sQuote(class(data_out)))),
+            "instead.",
+            "\n Check your inputs or contact app developer if error persists."
+          )
+        )
+      )
+
+
+      data_out
+    })
+
+    output$shiny_errors <- renderUI({
+      data_validated()
+      NULL
+    })
+
+    output$shiny_warnings <- renderUI({
+      if (inherits(data_out_rv(), "teal_data")) {
+        is_modules_ok <- check_modules_datanames(modules = modules, datanames = teal_data_ls(data_validated()))
+        if (!isTRUE(is_modules_ok)) {
+          tags$div(is_modules_ok, class = "teal-output-warning")
+        }
+      }
+    })
+
+    data_validated
+  })
+}
+
+#' Fallback on failure
+#'
+#' Function returns the previous reactive if the current reactive is invalid (throws error or returns NULL).
+#' Application: In `teal` we try to prevent the error from being thrown and instead we replace failing
+#' transform module data output with data input from the previous module (or from previous `teal` reactive
+#' tree elements).
+#'
+#' @param this (`reactive`) Current reactive.
+#' @param that (`reactive`) Previous reactive.
+#' @param label (`character`) Label for identifying problematic `teal_data_module` transform in logging.
+#' @return `reactive` `teal_data`
+#' @keywords internal
+.fallback_on_failure <- function(this, that, label) {
+  checkmate::assert_class(this, "reactive")
+  checkmate::assert_class(that, "reactive")
+  checkmate::assert_string(label)
+
+  reactive({
+    res <- try(this(), silent = TRUE)
+    if (inherits(res, "teal_data")) {
+      logger::log_debug("{ label } evaluated successfully.")
+      res
+    } else {
+      logger::log_debug("{ label } failed, falling back to previous data.")
+      that()
```

Comment:
Changed from `isolate(that())` to `that()` so that upstream data being modified will trigger this reactive.

Otherwise, this will return the first error fallback and not update as data from previous transforms (or when original from `data_rv()` `eventReactive` is in error)

This fixes the reactivity problem in #1303

Reply:
I will add a test for this scenario before we close the issue

Reply:
@averissimo please make sure that reactive never gets retriggered on `this()` and `that()` within the same reactive cycle. Should be easy to test 

<img width="644" alt="image" src="https://github.com/user-attachments/assets/b52952e5-dc9a-4edb-b264-081dde321c79">

Reply:
It seems to have no issues, I've tried a bunch of combinations and ended up with: c12b083

It changes the input at the same time so that `this()` no longer gives an error and `that()` has changed as well